lightly/api/api_workflow_datasources.py

```diff
@@ -0,0 +1,65 @@
+from typing import List
+from lightly.openapi_generated.swagger_client.models.datasource_raw_samples_data import (
+    DatasourceRawSamplesData,
+)
+from lightly.openapi_generated.swagger_client.models.datasource_raw_samples_data_row import (
+    DatasourceRawSamplesDataRow,
+)
+from lightly.openapi_generated.swagger_client.models.datasource_processed_until_timestamp_request import (
+    DatasourceProcessedUntilTimestampRequest,
+)
+
+import time
+
+def _convert_raw_data_to_list(data: List[DatasourceRawSamplesDataRow]):
+    return [(d.file_name, d.read_url) for d in data]
+
+
+class _DatasourcesMixin:
+
+    def download_raw_samples(self, _from: int = 0, to: int = None):
+        if to is None:
+            to = int(time.time())
+        response: DatasourceRawSamplesData = self._datasources_api.get_list_of_raw_samples_from_datasource_by_dataset_id(
+            dataset_id=self.dataset_id,
+            _from=_from,
+            to=to,
+        )
+        cursor = response.cursor
+        data = _convert_raw_data_to_list(response.data)
+        while response.has_more:
+            response: DatasourceRawSamplesData = self._datasources_api.get_list_of_raw_samples_from_datasource_by_dataset_id(
+                dataset_id=self.dataset_id, cursor=cursor
+            )
+            cursor = response.cursor
+            new_data = _convert_raw_data_to_list(response.data)
+            data.extend(new_data)
+        return data
```

Comment:
Several points here:
- `data` is a bad variable name
- What is the reason to change the `List[DatasourceRawSamplesDataRow]` to `List[Tuple[str, str]]`?
- The type hint for the return value is missing.
- What is the reason to `_convert_raw_data_to_list` both `data` and `new_data` instead of once data after the while loop? As this would call the function to only be called once, it could be inlined in this case and deleted.

Reply:
In general, I don't like returning `List[Tuple[str, str]]` over multiple instances, as you never know what the strings stand for. However, a `List[DatasourceRawSamplesDataRow]` is always fully clear, as you can use the intellisense to get the attribute names like `element.file_name` for each element.

Reply:
Alternatively, you could define a small class called `FileInfo`, with the `filename` and `url`. Or use a named tuple.

Reply:
If you have it as an explicit element, you also don't need to write down that the tuple must be (filename, url) in the docstrings of every function it is used.

Reply:
I do not return `List[DatasourceRawSamplesDataRow]` because I wanted to avoid having api objects in different parts of the codebase. And I did not use an extra class because I have not seen it used elsewhere in the code. For example all torch modules expect a batch to be a tuple of images, labels, and filenames without using an extra class. I completely agree that a class would make the code more readable. Another issue is that I am not sure if there is currently any module where it would make most sense to define such classes.